src/main/java/org/folio/circulation/domain/OverdueFineCalculatorService.java

```diff
@@ -67,12 +70,29 @@ private OverdueFineCalculatorService(Clients clients) {
   public CompletableFuture<Result<FeeFineAction>> createOverdueFineIfNecessary(
     CheckInProcessRecords records, String userId) {
 
-    Loan loan = records.getLoan();
+    return shouldChargeOverdueFineOnCheckIn(records)
+      .thenCompose(r -> r.after(shouldChargeOverdueFine -> {
+        if (!shouldChargeOverdueFine) {
+          return completedFuture(succeeded(null));
+        }
+        return createOverdueFineIfNecessary(records.getLoan(), Scenario.CHECKIN, userId);
+      }));
+  }
```

Comment:
I've raised a pull request (I got involved in trying to resolve it :-/) to change this in order to remove the code smell, please let me know your thoughts and apply it if you think it useful.

Reply:
Thank you, I've merged it.